opentelemetry/src/lib.rs

```diff
@@ -55,40 +35,109 @@
+//! customized in the [`opentelemetry_sdk`] based on your use case.
+//!
+//! ## Choosing the Right Instrument
+//! Selecting the correct instrument is critical for accurately representing
+//! your metrics data:
+//!
+//! - Use **Counters** for values that only increase, such as the number of
+//!   requests served or errors encountered.
+//! - Use **UpDownCounters** for values that can increase or decrease, such as
+//!   the number of active connections, number of items in a queue etc.
+//! - **Gauges:** Use for values that can go up or down and represent the
+//!   current state, such as CPU usage, temperature etc.
+//! - Use **Histograms** for measuring the distribution of a value, such as
+//!   response times or payload sizes.
+//!
+//! ### Observable Instruments
+//!
+//! Counters, UpDownCounters, and Gauges have Observable variants that allow
+//! values to be reported through a callback function. Observable instruments
+//! are ideal when the metric value is managed elsewhere and needs to be
+//! observed by OpenTelemetry instrumentation. The callbacks are automatically
+//! invoked by the OpenTelemetry SDK before every export (e.g., every 60
+//! seconds).
+//!
+//! For example:
+//! - An **ObservableCounter** can monitor the number of page faults in a
+//!   process as reported by the operating system.
+//! - An **ObservableUpDownCounter** can monitor the size of an in-memory queue
+//!   by reporting the size using queue's len() method within the callback
+//!   function.
+//!   
+//! For detailed guidance, refer to [OpenTelemetry Metrics API - Instrumentation
+//! Guidance](https://opentelemetry.io/docs/specs/otel/metrics/supplementary-guidelines/#instrument-selection).
+//!
+//! ## Best Practices
+//! - **Re-use Instruments:** Instruments are designed for
+//!   reuse. Avoid creating new instruments repeatedly.
+//! - **Clone for Sharing:** If the same instrument needs to be used across
+//!   multiple parts of your code, you can safely clone it to share.
+//!
+//! ## Example Usage
 //! ```
-//! # #[cfg(feature = "metrics")]
-//! # {
 //! use opentelemetry::{global, KeyValue};
 //!
-//! // get a meter from a provider
+//! // Get a meter from a provider.
 //! let meter = global::meter("my_service");
 //!
-//! // create an instrument
+//! // Create an instrument (in this case, a Counter).
 //! let counter = meter.u64_counter("my_counter").build();
 //!
-//! // record a measurement
+//! // Record a measurement by passing the value and a set of attributes.
 //! counter.add(1, &[KeyValue::new("http.client_ip", "83.164.160.102")]);
-//! # }
+//!
+//! // Create an ObservableCounter and register a callback that reports the measurement.
+//! let _observable_counter = meter
+//! .u64_observable_counter("my_observable_counter")
+//! .with_callback(|observer| {
+//!     observer.observe(
+//!         100,
+//!         &[
+//!             KeyValue::new("mykey1", "myvalue1"),
```

Comment:
It would be nice to stay consistent with the sample attributes: either use specific real-world examples (such as IP Address etc. used for Counter, Traces) or use generic examples such as `mykey1, myvalue1`. I would prefer using specific examples.